### ISSUE TITLE
Queues and scheduler rset not in sync

### ISSUE THREAD

**Issue by stonier** [utexas-bwi/concert_scheduling#15]
The scheduler will _grant_ a request, but in the request feedback back to the requester, the pre-grant status of the request is received (i.e. the grant update did not take effect).

I added some logging to get a handle on what is going on - the most interesting point of which is in the dispatch function. Passing it the rset variable via the callback method, i.e.

```
    def callback(self, rset):
        rospy.logdebug('scheduler callback:')
        for rq in rset.values():
            rospy.logwarn("DJS : request address in callback [%s]" % hex(id(rq)))
            rospy.logdebug('  ' + str(rq))
            if rq.msg.status == Request.NEW:
                self.queue(rq, rset.requester_id)
            elif rq.msg.status == Request.CANCELING:
                self.free(rq, rset.requester_id)
        self.dispatch(rset)                 # try to allocate ready requests

    def dispatch(self, rset=None):
        while len(self.ready_queue) > 0:
            # Try to allocate top element in the ready queue.
            elem = self.ready_queue.pop()
            rospy.logwarn("DJS: elem -> address [%s]" % hex(id(elem.request)))
            rospy.logwarn("DJS: elem -> id [%s]" % unique_id.toHexString(elem.request.msg.id))
            rospy.logwarn("DJS: elem -> status [%s]" % elem.request.msg.status)
            resources = []
            try:
                resources = self.pool.allocate(elem.request)
            except InvalidRequestError as ex:
                self.reject_request(elem, ex)
                continue                # skip to next queue element

            if not resources:           # top request cannot be satisfied?
                # Return it to head of queue.
                self.ready_queue.add(elem)
                break                   # stop looking

            try:
                elem.request.grant(resources)
                rospy.logwarn("DJS: dispatch -> address [%s]" % hex(id(elem.request)))
                rospy.logwarn("DJS: dispatch -> msg address [%s]" % hex(id(elem.request.msg)))
                rospy.logwarn("DJS: dispatch -> id [%s]" % unique_id.toHexString(elem.request.msg.id))
                rospy.logwarn("DJS: dispatch -> status [%s]" % elem.request.msg.status)
                if rset is not None:
                    for rq in rset.values():
                        rospy.logwarn("DJS: rq -> address [%s]" % hex(id(rq)))
                        rospy.logwarn("DJS: rq -> msg address [%s]" % hex(id(rq.msg)))
                        rospy.logwarn("DJS: rq -> id [%s]" % unique_id.toHexString(rq.msg.id))
                        rospy.logwarn("DJS: rq -> status [%s]" % rq.msg.status)
                rospy.loginfo(
                    'Request granted: ' + str(elem.request.uuid))
            except TransitionError:     # request no longer active?
                # Return allocated resources to the pool.
                self.pool.release_resources(resources)
            self.notification_set.add(elem.requester_id)
```

I get this output:

```
[WARN] [WallTime: 1396153452.730991] DJS : request address in callback [0x1f18490]
[WARN] [WallTime: 1396153452.731454] DJS: transitions [w] -> address [0x1f18490]
[WARN] [WallTime: 1396153452.731776] DJS: transitions [w] -> id [10d4e332-4e90-40f4-8f79-57c8bc098182]
[WARN] [WallTime: 1396153452.732043] DJS: transitions [w] -> grant [2]
[WARN] [WallTime: 1396153452.732398] DJS: transitions [w] -> resources [[rapp: rocon_apps/teleop
id: 
  uuid: [54, 104, 244, 114, 150, 196, 79, 31, 169, 128, 168, 115, 67, 248, 50, 146]
uri: rocon:/pc/guimul/hydro/precise
remappings: []]]
[INFO] [WallTime: 1396153452.733212] Request queued: 10d4e332-4e90-40f4-8f79-57c8bc098182
[WARN] [WallTime: 1396153452.733482] DJS: elem -> address [0x1f185d0]
[WARN] [WallTime: 1396153452.733764] DJS: elem -> id [10d4e332-4e90-40f4-8f79-57c8bc098182]
[WARN] [WallTime: 1396153452.734061] DJS: elem -> status [2]
[WARN] [WallTime: 1396153452.734742] DJS: transitions [g] -> address [0x1f185d0]
[WARN] [WallTime: 1396153452.735097] DJS: transitions [g] -> id [10d4e332-4e90-40f4-8f79-57c8bc098182]
[WARN] [WallTime: 1396153452.735369] DJS: transitions [g] -> grant [3]
[WARN] [WallTime: 1396153452.735695] DJS: transitions [g] -> resources [[rapp: rocon_apps/teleop
id: 
  uuid: [54, 104, 244, 114, 150, 196, 79, 31, 169, 128, 168, 115, 67, 248, 50, 146]
uri: rocon:/pc/guimul/hydro/precise
remappings: []]]
[WARN] [WallTime: 1396153452.736015] DJS: dispatch -> address [0x1f185d0]
[WARN] [WallTime: 1396153452.736313] DJS: dispatch -> msg address [0x1eedb40]
[WARN] [WallTime: 1396153452.736619] DJS: dispatch -> id [10d4e332-4e90-40f4-8f79-57c8bc098182]
[WARN] [WallTime: 1396153452.736906] DJS: dispatch -> status [3]
[WARN] [WallTime: 1396153452.737174] DJS: rq -> address [0x1f18490]
[WARN] [WallTime: 1396153452.737463] DJS: rq -> msg address [0x1eed830]
[WARN] [WallTime: 1396153452.737759] DJS: rq -> id [10d4e332-4e90-40f4-8f79-57c8bc098182]
[WARN] [WallTime: 1396153452.738022] DJS: rq -> status [2]
[INFO] [WallTime: 1396153452.738291] Request granted: 10d4e332-4e90-40f4-8f79-57c8bc098182
[WARN] [WallTime: 1396153452.738597] DJS: sending -> address [0x1f18490]
[WARN] [WallTime: 1396153452.738933] DJS: sending -> id 10d4e332-4e90-40f4-8f79-57c8bc098182
[WARN] [WallTime: 1396153452.739228] DJS: sending -> status 2
[WARN] [WallTime: 1396153452.739579] DJS: sending -> resources [rapp: rocon_apps/teleop
id: 
  uuid: [54, 104, 244, 114, 150, 196, 79, 31, 169, 128, 168, 115, 67, 248, 50, 146]
uri: rocon:/pc/guimul/hydro/precise
remappings: []]
[WARN] [WallTime: 1396153452.740037] DJS: resource pool changed
[WARN] [WallTime: 1396153452.740379] DJS: publishing known resources
[WARN] [WallTime: 1396153452.740843] DJS: requester feedback - request id 10d4e332-4e90-40f4-8f79-57c8bc098182
[WARN] [WallTime: 1396153452.741350] DJS: requester feedback - request status 2
[WARN] [WallTime: 1396153452.741736] DJS: requester feedback - request resources [rapp: rocon_apps/teleop
```

As you can see, the rset requets and the popped ready_queue request are not the same thing. So any `grant()` operation on the popped request is irrelevant. I suspect that they are supposed to be one and the same object?

**Comment by stonier** [comment]
Found the copy in [priority queue add()](https://github.com/utexas-bwi/rocon_experimental/blob/master/concert_simple_scheduler/src/concert_simple_scheduler/priority_queue.py#L100):

```
element = copy.deepcopy(element)
```

Question is now, whether the request object should be exactly the same as the object in the priority queue, or if it should be a copy and the grant process done back on the actual rset object, not the queue object in the dispatch function.

Are the rset request objects guaranteed to remain on a fixed memory address?

**Comment by stonier** [comment]
Commenting out that deepcopy works for my demos, but I worry it's volatile.

**Comment by jack-oquin** [comment]
I think we want the queue element to point to the original ActiveRequest object.

I am trying to figure out why I added a deep copy when adding it to the queue. Maybe it was just a mistake, but I must have had _something_ in mind.

**Comment by jack-oquin** [comment]
Commenting the deep copy out should be fine. 

I'll do some more testing and try to figure out the heck it was for in the first place.

**Comment by stonier** [comment]
ok.

**Comment by jack-oquin** [comment]
@stonier: I think 8674e20dea1dc722c6d07677695b97fe59de29f0 fixes the problem you detected. 

There _are_ reasons to copy the queue element. I believe that version is still broken when the request priority changes. That was the original reason for the deep copy. I need to think of a different solution for that.

We can live with it for a while, since our current system is fixed-priority, see #19.

**Comment by stonier** [comment]
Right, I'll give this a test shortly.

**Comment by jack-oquin** [comment]
I've got some ideas now to fix the priority queue properly, but that is a separate issue.

I think this one is fixed by 8674e20. The unit tests now assert that the original `ActiveRequest` instance must be retained.
